[3.x] Correctly support multiple events mapped to a single action

madmiraal opened the pull request:
3.x backport of #47599.

Although this breaks compat, the breaking changes may be considered worth it for 3.4. If not, it at least provides a way of testing the changes in #47599 using current 3.x projects.

The breaking changes are:
1. Calls to `InputEvent`'s `get_action_strength()`, `is_action_pressed()` and `is_action_released()` will now update the status of the action. Previously, only actual received events would.
~2. `InputMap`'s `event_is_action()` has been removed.~
3. `InputEventAction`'s strength property could be 0, while `is_pressed()` returned `true`.

I don't think any of them are critical, and the changes may actually be considered enhancements.
Note:
1. The `InputEvent` methods have duplicates in [`Input`](https://docs.godotengine.org/en/stable/classes/class_input.html), which I think make more sense to use. I would suggest removing the `InputEvent` versions, but they're used in the engine here:
https://github.com/godotengine/godot/blob/4a7679e4dd9f68270b3d27797146f88491f182b8/scene/main/viewport.cpp#L2399-L2430
2. `InputMap` also has `action_has_event()`, which is the same as `event_is_action()` with parameters reversed; except keyboard modifiers are not ignored.

~Finally, as with #47599, I've tested these changes with keyboard keys and mouse buttons, but I've added the `needs testing` label, because I don't have a gamepad to test `JoyButton` or `JoyAxis` events, or confirm that it fixes #39287 and #45628.~

**Edit:** I have now personally tested this with a controller to confirm that it does fix #39287 and #45628, and reverted the deletion of `event_is_action()` to minimise compat breaking.
**Note:** It now includes automatically updating projects' input settings (i.e. it updates a project's `project.godot` file including the `config_version`). This means that projects opened with this PR cannot be opened with an earlier version of Godot (without restoring the `project.godot` file).

madmiraal committed:
Correctly support multiple events mapped to a single action

Stores the state of each event to enable determining the correct state
of the action regardless of which event is used. This ensures that when
two or more events are mapped to a single action, the Input
is_action_pressed() reports true if any of the events are still pressed,
and get_action_strength() returns the maximum value provided by all the
events.

Automatically updates projects' input settings.

mphe commented:
I just tested this fix on the 3.x branch using the reproduction project from #45628 and a PS3 controller + keyboard. Unfortunetly it is still broken when an action contains a controller axis. Without a controller axis, keys seem to work fine.

The results are:
* Keys sometimes get "stuck" and still return 1 endlessly after the key was released.
* Keys sometimes don't return 1 at all when pressed.
* Controller axes always return 0, but using them resets the key bugs above.

madmiraal commented:
Updated to fix issues identified by @mphe, and rebased to incorporate the changes made to `Input` since this PR was created. It now includes automatically updating projects' input settings (i.e. it updates a project's `project.godot` file including the `config_version`). This means that projects opened with this PR cannot be opened with an earlier version of Godot (without restoring the `project.godot` file).

I have now personally tested this with a controller to confirm that it does fix #39287 and #45628.

superoven reviewed (commented):
^ Minor nitpick I noticed.

I tried this on my project where I was experiencing this issue with a single input definition having both a keyboard and joystick mapping causes issues. FWIW, this PR fixes the issue for me.

madmiraal commented on core/os/input_event.h:
```diff
@@ -215,18 +223,20 @@ class InputEvent : public Resource {
 
-	// To be removed someday, since they do not make sense for all events
-	virtual bool is_pressed() const;
+	// To be removed someday, since it doesn't make sense for all events
```
The comment referred to both `is_pressed()` and `is_echo()` and now it refers to only `is_echo()`; hence the change in the wording from "they" to "it".
